### PR TITLE
Generating XML with kept ordering.

### PR DESCRIPTION
**Pull request by greut**
- A record is filled into a `GroupOrderedDict`. An immutable object that keeps things in order.
- MARC XML can be generated too.
- MARC JSON format has a key called `__order__` to be able to reconstruct it.

### CODE REVIEW

**Comment by greut** [comment]
**Problems**
Two behaviours for the `GroupableOrderedDict`
1. everything it returns is a `tuple`.
2. values it returns are either a `tuple` of values or the value itself if one.

---

**Issues with solution1:** `schemas/bdxxxx.py` does `utils.force_list()` on expected list items but doesn't undo any non-list item.

``` python
# Here the access_method is a lone field
{'electronic_location_and_access': ['access_method': ('Awesome access method',), ...]}
```

`access_method` is not a `force_list` field but as the data structure returns a list, it takes the list.

---

**Issues with solution 2:** `to_marc21.do` will put things in lists that aren't the in the `JSON` or `GroupableOrderedDict`. Pretty hard to test.

``` python
{'a': 'https://github.com/bcmd/bphi/tree/v1.0',
 'i': ('isSupplementTo',),
 'n': ('url',)},
```

The original data structure will not have the list around `i` or `n` because it has no clue about its structure.

File: dojson/contrib/marc21/utils.py [hunk 1]
```diff
@@ -24,6 +23,140 @@
...
+        try:
+            return self.__getitem__(key)
+        except KeyError:
+            return default
+
+    def __getitem__(self, key):
+        if key == "__keys__":
+            return self.__keys__
+        if key in self._keys:
+            return tuple([v for k, v in self._data if k == key])
```
Comment (jirikuncar) [inline dojson/contrib/marc21/utils.py]: do we need tuple?

Comment (greut) [inline dojson/contrib/marc21/utils.py]: nope, but it shows the immutabliity of the data structure.

**Comment by jirikuncar** [comment]
### Possible data flow
- load: source in MARC XML -> `load+create_record` -> "recstruct" of type `GroupableOrderedDict` -> `marc21.do` -> "JSONable" dict with `__order__` keys
- dump: "JSONable" dict with `__order__` keys -> `to_marc21.do` (translate `__order__` keys) -> "recstruct" of type `GroupableOrderedDict` -> generate MARC XML with `__order__` in mind

**Comment by greut** [comment]
@jirikuncar I'll go with if the key has one element, return it as is rather than as a list because `marc21.do` has no `utils.force_item`.

**Comment by jirikuncar** [comment]
@greut yes. It will behave better with non-repeatable (sub)fields. Thanks

File: dojson/contrib/marc21/utils.py [hunk 2]
```diff
@@ -24,6 +23,140 @@
...
+        for k, v in self.iteritems(with_order=False):
+            if len(v) != len(other[k]):
+                return False
+
+            for i, value in enumerate(v):
+                if other[k][i] != value:
+                    return False
+        return True
+
+    def get(self, key, default=None):
```
Comment (jirikuncar) [inline dojson/contrib/marc21/utils.py:107]: IMHO we don't need it as `__getitem__` is implemented.

File: dojson/contrib/marc21/utils.py [hunk 3]
```diff
@@ -42,84 +175,40 @@ def create_record(marcxml, correct=False, keep_singletons=True):
...
-                    (key + value[1] + value[2]).replace(' ', '_'), field)
-
-    return rec_tree
+                fields.append((code, text))
+
+        if fields or keep_singletons:
+            key = '{0}{1}{2}'.format(tag, ind1, ind2)
+            record.append((key, GroupableOrderedDict(fields)))
+
+    return OrderedDict(record)
```
Comment (jirikuncar) [inline dojson/contrib/marc21/utils.py]: I'm not sure if it's going to work for repeatable fields with same indicators. IMHO we should use `GroupableOrderedDict`. If there is no test-case I will try to provide one.

**Comment by jirikuncar** [comment]
Test-case data:

``` python
RECORD_REPEATABLE_FIELDS = """<record>
  <datafield tag="100" ind1=" " ind2=" ">
    <subfield code="a">Donges, Jonathan F</subfield>
  </datafield>
  <datafield tag="700" ind1="1" ind2=" ">
    <subfield code="a">Jefferson, Thomas</subfield>
    <subfield code="d">1743-1826</subfield>
    <subfield code="e">former owner</subfield>
    <subfield code="5">MH</subfield>
  </datafield>
  <datafield tag="700" ind1="1" ind2=" ">
    <subfield code="a">Ford, John</subfield>
    <subfield code="d">1894-1973</subfield>
    <subfield code="e">direction</subfield>
  </datafield>
  <datafield tag="700" ind1="0" ind2=" ">
    <subfield code="a">Father Divine</subfield>
  </datafield>
</record>"""
```

File: dojson/contrib/marc21/utils.py (hunk 2 above)
Comment (greut) [inline dojson/contrib/marc21/utils.py:107]: It doesn't seem to work. This is a `dict` and not a `Mapping` or `MutableMapping` anymore.

Comment (jirikuncar) [inline dojson/contrib/marc21/utils.py:107]: Can we keep the `Mapping` parent or is there a drawback?

**Comment by greut** [comment]
@jirikuncar Thanks!

**Comment by greut** [comment]
The tests are passing but the manual fix on the controlfields feels wrong.

File: dojson/contrib/marc21/fields/bd00x.py [hunk 4]
```diff
@@ -15,7 +15,7 @@
 @marc21.over('control_number', '^001')
 def control_number(self, key, value):
     """Control Number."""
-    return value[0]
+    return value
```
Comment (jirikuncar) [inline dojson/contrib/marc21/fields/bd00x.py]: it should be changed in all 00x fields

**Comment by greut** [comment]
reorder, stash, squash, unsquash, edit, amend, cut, paste, etc.

**Pull request merged**